### ISSUE TITLE
Why do I fail to duplicate AdventureWorks2019 with scaffolding followed by migration?

### ISSUE THREAD

**Issue by pstricks-fans** [dotnet/efcore#27644]
## Objective

For learning purposes, I want to make a new copy of unpopulated database from the existing populated one through the following steps:
- do reverse-enginering on the existing database (with `scaffold-dbcontext`) to get its entity models and database context,
- do migration (with `add-migration`) and apply it to generate the brand new unpopulated database.

## Setup

My simulating project uses

- `AdventureWorks2019.bak` (OLTP version) downloaded from [Microsoft](https://docs.microsoft.com/en-us/sql/samples/adventureworks-install-configure?view=sql-server-ver15&tabs=ssms).

- Two connection strings.
  ```  
  "ConnectionStrings": {
    "Original": "server=localhost;initial catalog=OriginalDb;trusted_connection=true",
    "Duplicate": "server=localhost;initial catalog=DuplicateDb;trusted_connection=true"
  }
  ```

- a host.
    ```
    class Program
    {
        static void Main()
        {
            Host.CreateDefaultBuilder()
                .ConfigureServices((ctx, isc) =>
                {
                    // Models.TheContext is available for migration 
                    // only after invoking scaffold-dbcontext

                    //isc.AddDbContext<Models.TheContext>(options =>
                    //{
                    //    options.UseSqlServer(ctx.Configuration.GetConnectionString("Duplicate"));
                    //});
                })
                .Build();
        }
    }
    ```

My complete repo can be found [here](https://github.com/pstricks-fans/ScaffoldAndCreateAdventureWorks). 

## How to reproduce

 - Restore `AdventureWorks2019.bak` and name the restored database `OriginalDb` (to match against one of my connection strings named `Original`).
 
 - Compile the project and do reverse-engineering on the existing database with 
   ```
    scaffold-dbcontext name=Original microsoft.entityframeworkcore.sqlserver -context TheContext -NoOnConfiguring -OutputDir Models
   ```

    I get the following warnings

    ```
    // Unable to generate entity type for table 'Production.Document' since its primary key could not be scaffolded. Please see the warning messages.
    // Unable to generate entity type for table 'Production.ProductDocument' since its primary key could not be scaffolded. Please see the warning messages.
    ```

    and some classes in `Models` folder.

- Uncomment out the part using `Models.TheContext` and recompile the project. Afterwards, do migration with
   ```
   add-migration Initialization
   ```
   I get some (trimmed) warnings 

   ```
    Microsoft.EntityFrameworkCore.Model.Validation[30001]
      The property 'SystemInformationId' on entity type 'AwbuildVersion' is of type 'byte', but is set up to use a SQL Server identity column; this requires that values starting at 255 and counting down will be used for temporary key values. A temporary key value is needed for every entity inserted in a single call to 'SaveChanges'. Care must be taken that these values do not collide with real key values.
    Microsoft.EntityFrameworkCore.Model.Validation[30001]
      The property 'ShiftId' on entity type 'Shift' is of type 'byte', but is set up to use a SQL Server identity column; this requires that values starting at 255 and counting down will be used for temporary key values. A temporary key value is needed for every entity inserted in a single call to 'SaveChanges'. Care must be taken that these values do not collide with real key values.
    Microsoft.EntityFrameworkCore.Infrastructure[10403]
    ```

   And when I execute 
   ```
   update-database
   ```
   I get the following (trimmed) errors and the generated database only has  one table (the migration history table).
   ```
      Failed executing DbCommand (5ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
		CREATE TABLE [HumanResources].[Employee] (
		[BusinessEntityID] int NOT NULL,
		[NationalIDNumber] nvarchar(15) NOT NULL,
		[ModifiedDate] datetime NOT NULL DEFAULT ((getdate())),
		CONSTRAINT [PK_Employee_BusinessEntityID] PRIMARY KEY ([BusinessEntityID]),
		CONSTRAINT [FK_Employee_Person_BusinessEntityID] FOREIGN KEY ([BusinessEntityID]) REFERENCES [Person].[Person] ([BusinessEntityID])
		);
		DECLARE @description AS sql_variant;
		SET @description = N'Employee information such as salary, department, and title.';
		EXEC sp_addextendedproperty 'MS_Description', @description, 'SCHEMA', N'HumanResources', 'TABLE', N'Employee';
		SET @description = N'Primary key for Employee records.  Foreign key to BusinessEntity.BusinessEntityID.';
		EXEC sp_addextendedproperty 'MS_Description', @description, 'SCHEMA', N'HumanResources', 'TABLE', N'Employee', 'COLUMN', N'BusinessEntityID';
		SET @description = N'Unique national identification number such as a social security number.';
		EXEC sp_addextendedproperty 'MS_Description', @description, 'SCHEMA', N'HumanResources', 'TABLE', N'Employee', 'COLUMN', N'NationalIDNumber';
		SET @description = N'Network login.';
		EXEC sp_addextendedproperty 'MS_Description', @description, 'SCHEMA', N'HumanResources', 'TABLE', N'Employee', 'COLUMN', N'LoginID';

		EXEC sp_addextendedproperty 'MS_Description', @description, 'SCHEMA', N'HumanResources', 'TABLE', N'Employee', 'COLUMN', N'HireDate';
		SET @description = N'Job classification. 0 = Hourly, not exempt from collective bargaining. 1 = Salaried, exempt from collective bargaining.';
		EXEC sp_addextendedproperty 'MS_Description', @description, 'SCHEMA', N'HumanResources', 'TABLE', N'Employee', 'COLUMN', N'SalariedFlag';
		SET @description = N'Number of available vacation hours.';
		EXEC sp_addextendedproperty 'MS_Description', @description, 'SCHEMA', N'HumanResources', 'TABLE', N'Employee', 'COLUMN', N'VacationHours';

		SET @description = N'Date and time the record was last updated.';
		EXEC sp_addextendedproperty 'MS_Description', @description, 'SCHEMA', N'HumanResources', 'TABLE', N'Employee', 'COLUMN', N'ModifiedDate';
		Cannot find either column "OrganizationNode" or the user-defined function or aggregate "OrganizationNode.GetLevel", or the name is ambiguous.
   ```

## Question

What am I wrong here and how to fix it?

**Comment by Regenhardt** [comment]
What warnings do you get during the first step, scaffolding?

**Comment by pstricks-fans** [comment]
> // Unable to generate entity type for table 'Production.Document' since its primary key could not be scaffolded. Please see the warning messages.
> // Unable to generate entity type for table 'Production.ProductDocument' since its primary key could not be scaffolded. Please see the warning messages.

I already provided the warnings:

```
// Unable to generate entity type for table 'Production.Document' since its primary key could not be scaffolded. Please see the warning messages.
// Unable to generate entity type for table 'Production.ProductDocument' since its primary key could not be scaffolded. Please see the warning messages.
```

**Comment by pstricks-fans** [comment]
> What warnings do you get during the first step, scaffolding?
I already provided the warnings:

```
// Unable to generate entity type for table 'Production.Document' since its primary key could not be scaffolded. Please see the warning messages.
// Unable to generate entity type for table 'Production.ProductDocument' since its primary key could not be scaffolded. Please see the warning messages.
```

**Comment by Regenhardt** [comment]
What does the "Please see the warning messages" refer to though? I'd expect some more warnings somewhere in the log.

**Comment by pstricks-fans** [comment]
> What does the "Please see the warning messages" refer to though? I'd expect some more warnings somewhere in the log.

No more warning messages are available. There are only two  lines as I quoted above.

**Comment by ajcvickers** [comment]
@pstricks-fans AdventureWorks is designed to use as many features of SQL Server has possible. Not all of these features are supported by EF Core.

Beyond that, it is not a goal of scaffolding and migrations that they be completely round-tripable. This is because both directions are lossy. That is, the database may contain things that cannot be represented in the EF model, and the EF model may contain things that cannot be represented in the database.

**Comment by pstricks-fans** [comment]
> @pstricks-fans AdventureWorks is designed to use as many features of SQL Server has possible. Not all of these features are supported by EF Core.
> 
> Beyond that, it is not a goal of scaffolding and migrations that they be completely round-tripable. This is because both directions are lossy. That is, the database may contain things that cannot be represented in the EF model, and the EF model may contain things that cannot be represented in the database.

Thank you very much for your response. OK. I close this issue.
